hrishikeshpaul/vibey
# Feat/sm/refresh token

***What:***
* Adds the access token and refresh token flow to login
* Adds valid tokens to "whitelist"
* Removes tokens from "whitelist" on log-out (essentially blacklisting them)
* Adds token validation middleware (validates and refreshes if necessary)
* Adds a generic error handler class and utility function to organize consistent error handling

***Why:***
* to add login/logout functionality and the required middleware for routes that requires authorization

***Testing this PR***:
* we need to write unit tests for sure, but things are sort of complicated with handling jwt in unit tests, I'll try to work on it soon unless you'd like to.  In the meantime, here's how I think we can manually test it:
* Sign into vibey and open up postman on the side
* Copy your access token and refresh token over to postman
* Locally, add the isLoggedIn middleware to the `/logout` route
* modify your refresh token on postman
* Try to hit the logout route on postman, you should receive a 403 error (i think 403)
* correct the refresh token in postman
* hit the route again, it should successfully log out
* try to hit the route again, it should send an error saying invalid tokens or something of that sort (meaning the tokens are now removed from the whitelist)

***Limitations/issues:***
* AT and RT are currently always being passed together, so it's debatable whether two tokens is necessary at the moment (perhaps two tokens is just more secure?).  With that being said, perhaps we could change things around a bit in the future
